kjdoc/tour.md

```diff
@@ -538,13 +538,440 @@ This section describes KJ APIs that control process execution and low-level inte
+
+You may use `promise.attach(kj::mv(object))` to give a `Promise` direct ownership of an object that must be kept alive until the promise completes. `.attach()`, like `.then()`, consumes the promise and returns a new one of the same type.
+
+```c++
+kj::Promise<kj::Own<kj::AsyncIoStream>> connectPromise =
+    networkAddress.connect();
+kj::Promise<kj::String> textPromise = connectPromise
+    .then([](kj::Own<kj::AsyncIoStream> stream) {
+  // We must attach the stream so that it remains alive until `readAllText()`
+  // is done. The stream will then be discarded.
+  return stream->readAllText().attach(kj::mv(stream));
+});
+```
+
+Using `.attach()` is semantically equivalent to using `.then()`, passing an identify function as the continuation, while having that function capture ownership of the attached object, i.e.:
+
+```c++
+// This...
+promise.attach(kj::mv(attachment));
+// ... is equivalent to this...
+promise.then([a = kj::mv(attachment)](auto x) { return kj::mv(x); });
+```
+
+Note that you can use `.attach()` together with `kj::defer()` to construct a "finally" block -- code which will execute after the promise completes (or is canceled).
+
+```c++
+promise = promise.attach(kj::defer([]() {
+  // This code will execute when the promise completes or is canceled.
+}));
+```
+
+
+### Background tasks
+
+If you construct a `Promise` and then just leave it be without calling `.then()` or `.wait()` to consume it, the task it represents will execute in the background. You can call `.then()` or `.wait()` later on, when you're ready. This makes it possible to run multiple concurrent tasks at once.
+
+Note that, when possible, KJ evaluates continuations lazily. Continuations which merely transform the result (without returning a new `Promise` that might require more waiting) are only evaluated when the final result is actually needed. This is an optimization which allows a long chain of `.then()`s to be executed all at once, rather than turning the event loop for each one. However, it can lead to some confusion when storing an unconsumed `Promise`. For example:
+
+```c++
+kj::Promise<void> timer.afterDelay(5 * kj::SECONDS)
+    .then([]() {
+  // This log line will never be written, because nothing
+  // is waiting on the final result of the promise.
+  KJ_LOG(WARNING, "It has been 5 seconds!!!");
+});
+kj::NEVER_DONE.wait(waitScope);
+```
+
+To solve this, use `.eagerlyEvaluate()`:
+
+```c++
+kj::Promise<void> timer.afterDelay(5 * kj::SECONDS)
+    .then([]() {
+  // This log will correctly be written after 5 seconds.
+  KJ_LOG(WARNING, "It has been 5 seconds!!!");
+}).eagerlyEvaluate([](kj::Exception&& exception) {
+  KJ_LOG(ERROR, exception);
+});
+kj::NEVER_DONE.wait(waitScope);
```

Comment:
I suppose it might be possible for someone to skim these examples and conclude that `.eagerlyEvaluate()` somehow makes the promise run synchronously.  Not sure if there's an easy way to dispel this.

Reply:
Yeah, I can't think of anything to do about this... people will have to read the surrounding text.